### PR TITLE
arch/irq: add the up_getusrpc macro to get the PC of the interrupted …

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary

arch/irq: add the up_getusrpc macro 
to get the PC of the interrupted thread in the interrupt

## Impact

none, new API

## Testing

bes board

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



**No, this PR does not meet the NuttX requirements.** 

Here's why and how to improve it:

**Missing Information:**

* **Summary:**
    * **Why is this change necessary?**  What problem does it solve? Why is retrieving the PC of the interrupted thread from an interrupt handler useful? 
    * **How does it work?** Provide a more detailed explanation of the `up_getusrpc` macro. What does it do internally? Are there any limitations or special considerations?
* **Impact:**
    * **"none, new API" is not sufficient.**  Even new APIs can have potential impacts:
        *  **Compatibility:** Could this API conflict with existing or future code?
        *  **Security:** Are there any security implications of exposing this information in an interrupt context?
        * **Documentation:** New APIs *always* require documentation.
* **Testing:**
    * **Insufficient detail:** 
        * "bes board" doesn't specify the architecture, board configuration, or NuttX version. 
        * Provide the full board name (e.g., "stm32f4discovery").
    * **No logs:** Provide actual testing logs demonstrating the change's functionality:
        * Show how to use the `up_getusrpc` macro.
        * Demonstrate the output or behavior before and after the change.

**How to Improve the PR:**

1. **Expand the Summary:** Clearly explain the problem, the solution, and how it works.
2. **Thoroughly Address Impact:** Consider all potential consequences, even for seemingly small changes.
3. **Provide Detailed Testing Information:**  Include:
    * Build host details (OS, CPU, compiler).
    * Target details (architecture, board, NuttX version/commit).
    *  **Complete testing logs** showing the macro's usage and the results before and after the change. 

By providing this information, you make it easier for reviewers to understand and evaluate your changes, increasing the chances of your PR being accepted.

File: arch/avr/include/avr/irq.h [hunk 1]
```diff
@@ -203,6 +203,21 @@ static inline void up_irq_restore(irqstate_t flags)
  * Public Function Prototypes
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: up_getusrpc
+ ****************************************************************************/
+
+#if defined(REG_PC2)
+#  define up_getusrpc(regs) \
+    ((((uint8_t *)((regs) ? (regs) : g_current_regs))[REG_PC0] << 16) | \
```
Comment (anchao) [inline arch/avr/include/avr/irq.h:212]:
```suggestion
    ((((uint8_t *)((regs != NULL) ? (regs) : CURRENT_REGS))[REG_PC0] << 16) | \
```

Comment (xiaoxiang781216) [reply arch/avr/include/avr/irq.h:212]: CURRENT_REGS doesn't exist anymore， should use up_current_reg instead.

Comment (anchao) [reply arch/avr/include/avr/irq.h:212]: If so, why other architectures still use CURRENT_REGS

File: arch/avr/include/avr/irq.h [hunk 2]
```diff
@@ -203,6 +203,21 @@ static inline void up_irq_restore(irqstate_t flags)
...
 
+/****************************************************************************
+ * Name: up_getusrpc
+ ****************************************************************************/
+
+#if defined(REG_PC2)
+#  define up_getusrpc(regs) \
+    ((((uint8_t *)((regs) ? (regs) : up_current_regs()))[REG_PC0] << 16) | \
+     (((uint8_t *)((regs) ? (regs) : up_current_regs()))[REG_PC1] <<  8) | \
+     (((uint8_t *)((regs) ? (regs) : up_current_regs()))[REG_PC2] <<  0))
```
Comment (anchao) [inline arch/avr/include/avr/irq.h:214]: check the regs before shift value

File: arch/avr/include/avr/irq.h [hunk 3]
```diff
@@ -203,6 +203,21 @@ static inline void up_irq_restore(irqstate_t flags)
...
+#else
+#  define up_getusrpc(regs) \
+    ((((uint8_t *)((regs) ? (regs) : up_current_regs()))[REG_PC0] << 8) | \
+     (((uint8_t *)((regs) ? (regs) : up_current_regs()))[REG_PC1] << 0))
```
Comment (anchao) [inline arch/avr/include/avr/irq.h:218]: ditto

**Review by anchao — CHANGES_REQUESTED** [review]

**Pull request merged**